File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        answer = [1] * n
        
        # Calculate prefix products and store in answer
        prefix_product = 1
        for i in range(n):
            answer[i] = prefix_product
            prefix_product *= nums[i]
        
        # Calculate suffix products and multiply with prefix products in answer
        suffix_product = 1
        for i in range(n - 1, -1, -1):
            answer[i] *= suffix_product
            suffix_product *= nums[i]
        
        return answer

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.ts
function productExceptSelf(nums: number[]): number[] {
    let products = Array(nums.length).fill(1);
    let prefixProduct = 1;
    for (let i = 0; i < nums.length; i++) {
        products[i] *= prefixProduct
        prefixProduct *= nums[i]
    }
    let suffixProduct = 1;
    for (let i = nums.length - 1; i >= 0; i--) {
        products[i] *= suffixProduct
        suffixProduct *= nums[i]
    }
    return products;
};